### PR TITLE
doc: add pandas and xarray fixtures to testing API docs

### PR DESCRIPTION
**Pull request by story645**
## PR summary

Motivated by #28830, figured it'd be a lot easier for contributors to use/reviewers to suggest if the pandas and xarray fixtures were exposed in the docs. I'm not fond of the subheading title but also can't think of a better one.  

## PR checklist

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by timhoffm** [comment]
Does not render for some reason.

**Comment by timhoffm** [comment]
That's what I meant. For some reason autofunction does not produce content here.

If nothing else helps, you could resort to a manual `.. Py::function::`. Not preferred to have doc content in rst away from the function, IMHO bearable in this special case.

**Comment by timhoffm** [comment]
Theory: I suspect the reason is that `matplotlib.testing.conftest` imports `pytest`. This fails because `pytest` is not present in the doc build environment. I know that Sphinx autodoc silently ignores failed package imports and does not produce output.

You may try [autodoc_mock_imports](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_mock_imports) but I don't know whether that can mock decorators.

File: doc/api/testing_api.rst [hunk 1]
```diff
@@ -37,3 +37,10 @@
    :members:
    :undoc-members:
    :show-inheritance:
+
+
+Data container testing fixtures
+==================================
```
Comment (timhoffm) [inline doc/api/testing_api.rst:43]: How about

```suggestion
Testing with optional dependencies
==================================
```

**Review by timhoffm — COMMENTED** [review]
Mock import works 👍 

> Was wondering if I should maybe add sample usage to the docstrings

This would be very helpful.

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
@story645 you can add the usage examples or merge immediately - as you see fit.

File: lib/matplotlib/testing/conftest.py [hunk 2]
```diff
@@ -82,7 +82,20 @@ def mpl_test_settings(request):
 
 @pytest.fixture
 def pd():
-    """Fixture to import and configure pandas."""
+    """
+    Fixture to import and configure pandas.
+
+    Example
+    -------
```
Comment (timhoffm) [inline lib/matplotlib/testing/conftest.py:89]: Try this:
```suggestion
    Examples
    --------
```

File: lib/matplotlib/testing/conftest.py [hunk 3]
```diff
@@ -95,6 +108,20 @@ def pd():
 
 @pytest.fixture
 def xr():
-    """Fixture to import xarray."""
+    """
+    Fixture to import xarray.
+
+    Example
+    -------
```
Comment (timhoffm) [inline lib/matplotlib/testing/conftest.py:115]:
```suggestion
    Examples
    --------
```

**Review by timhoffm — COMMENTED** [review]
"Example" is not a numpydoc section. It's the plural https://numpydoc.readthedocs.io/en/latest/format.html#examples

Maybe this is tripping up numpydoc/autodoc?

File: lib/matplotlib/testing/conftest.py [hunk 4]
```diff
@@ -82,7 +82,20 @@ def mpl_test_settings(request):
 
 @pytest.fixture
 def pd():
-    """Fixture to import and configure pandas."""
+    """
+    Fixture to import and configure pandas.
```
Comment (QuLogic) [inline lib/matplotlib/testing/conftest.py:86]: This should mention that it makes pandas optional.

File: lib/matplotlib/testing/conftest.py [hunk 5]
```diff
@@ -95,6 +108,20 @@ def pd():
 
 @pytest.fixture
 def xr():
-    """Fixture to import xarray."""
+    """
+    Fixture to import xarray.
```
Comment (QuLogic) [inline lib/matplotlib/testing/conftest.py:112]: Ditto.

File: lib/matplotlib/testing/conftest.py (hunk 4 above)
Comment (timhoffm) [reply lib/matplotlib/testing/conftest.py:86]:
```suggestion
    Fixture to import and configure the optional dependency pandas.
    
    The fixture is basically equal to `pd = pytest.importorskip('pandas')`
    plus some additional configuration.
    
    Use this whenever you need the `pandas` namespace in tests.
```

**Comment by QuLogic** [comment]
Well, it currently outputs a 27K-line table of issues on just the files in `lib/matplotlib/*.py`, so I'm not sure how feasible that is right now.

File: lib/matplotlib/testing/conftest.py [hunk 6]
```diff
@@ -82,7 +82,20 @@ def mpl_test_settings(request):
...
 @pytest.fixture
 def pd():
-    """Fixture to import and configure pandas."""
+    """
+    Fixture to import and configure pandas so that test is skipped when pandas is not
+    installed.
+
+    Examples
+    --------
+    Request the pandas fixture by passing in `pd` as an argument to the test ::
```
Comment (timhoffm) [inline lib/matplotlib/testing/conftest.py:91]:
```suggestion
    Request the pandas fixture by passing in ``pd`` as an argument to the test ::
```

**Comment by timhoffm** [comment]
> I know linting fix PRs aren't folks faves, but this seems like a good low hanging fruit task for sprints

Fine as long as the resulting PRs are trivial to review (small enough, only one kind of change at a time).

I suspect the 27k-lines of issues contain a lot of aspects that we cannot or don't want to fulfill. I think you'd have to start with a `checks` whitelist to work your way up and not get overwhelmed with messages. Then, when the big chunks are done, switch to the blacklist and narrow down further.

**Pull request merged**